pkg/sync/syncWorker.go

```diff
@@ -0,0 +1,205 @@
+package sync
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/xmtp/xmtpd/pkg/db/queries"
+	"github.com/xmtp/xmtpd/pkg/proto/xmtpv4/message_api"
+	"github.com/xmtp/xmtpd/pkg/registrant"
+	"github.com/xmtp/xmtpd/pkg/registry"
+	"github.com/xmtp/xmtpd/pkg/tracing"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/protobuf/proto"
+)
+
+type syncWorker struct {
+	ctx          context.Context
+	log          *zap.Logger
+	nodeRegistry registry.NodeRegistry
+	registrant   *registrant.Registrant
+	store        *sql.DB
+	wg           sync.WaitGroup
+}
+
+func startSyncWorker(
+	ctx context.Context,
+	log *zap.Logger,
+	nodeRegistry registry.NodeRegistry,
+	registrant *registrant.Registrant,
+	store *sql.DB,
+) (*syncWorker, error) {
+	s := &syncWorker{
+		ctx:          ctx,
+		log:          log.With(zap.String("method", "syncWorker")),
+		nodeRegistry: nodeRegistry,
+		registrant:   registrant,
+		store:        store,
+		wg:           sync.WaitGroup{},
+	}
+	if err := s.start(); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func (s *syncWorker) start() error {
+	nodes, err := s.nodeRegistry.GetNodes()
+	if err != nil {
+		return err
+	}
+	for _, node := range nodes {
+		if node.NodeID == s.registrant.NodeID() || !node.IsHealthy {
+			continue
+		}
+		s.subscribeToNode(node)
+	}
+	return nil
+}
+
+func (s *syncWorker) close() {
+	s.wg.Wait()
+}
+
+func (s *syncWorker) subscribeToNode(node registry.Node) {
+	tracing.GoPanicWrap(
+		s.ctx,
+		&s.wg,
+		fmt.Sprintf("node-subscribe-%d", node.NodeID),
+		func(ctx context.Context) {
+			var err error
+			for {
+				if err != nil {
+					log.Error(fmt.Sprintf("Error: %v, retrying...", err))
+					time.Sleep(1 * time.Second)
+				}
+				err = nil
+
+				conn, err := s.connectToNode(node)
+				if err != nil {
+					continue
+				}
+				stream, err := s.setupStream(node, conn)
+				if err != nil {
+					continue
+				}
+				err = s.listenToStream(stream)
+				if err != nil {
+					continue
+				}
+			}
+		},
+	)
+}
+
+func (s *syncWorker) connectToNode(node registry.Node) (*grpc.ClientConn, error) {
+	addr := node.HttpAddress
+	log.Info(fmt.Sprintf("attempting to connect to %s", addr))
+	conn, err := grpc.NewClient(
+		addr,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
```

Comment:
Will this work with TLS in the real environment?

Reply:
This brings up a good point. We don't have a standard way to specify if the server expects a TLS connection or not. In deployed environments we want to make sure that all servers connect via TLS, but it's a pain to enforce that locally.

We could assume that urls prefixed with `https://` are TLS and anything else isn't. Or we could add more metadata to the node registry.

Reply:
Great catch, I added a work item `Allow registry to specify TLS endpoints for prod and non-TLS endpoints for dev` for now. We should also figure out what format we want the URL's to be in - right now we're setting them to a GRPC-specific format (prefixed with `dns://` or `passthrough://` rather than `http://` and https://`